receiver/kafkareceiver/consumer_franz.go

```diff
@@ -0,0 +1,475 @@
+		ReceiverCreateSettings: c.settings,
+	})
+	if err != nil {
+		return err
+	}
+	c.obsrecv = obsrecv
+
+	// Create franz-go consumer client
+	client, err := kafka.NewFranzConsumerGroup(ctx,
+		c.config.ClientConfig,
+		c.config.ConsumerConfig,
+		c.topics,
+		c.settings.Logger,
+		kgo.OnPartitionsAssigned(c.assigned),
+		kgo.OnPartitionsRevoked(func(ctx context.Context, _ *kgo.Client, m map[string][]int32) {
+			c.lost(ctx, c.client, m, false)
+		}),
+		kgo.OnPartitionsLost(func(ctx context.Context, _ *kgo.Client, m map[string][]int32) {
+			c.lost(ctx, c.client, m, true)
+		}),
+	)
+	if err != nil {
+		return err
+	}
+	c.client = client
+
+	cm, err := c.newConsumeFn(host, c.obsrecv, c.telemetryBuilder)
+	if err != nil {
+		return err
+	}
+	c.consumeMessage = cm
+
+	go c.consumeLoop(c.processingCtx)
+	return nil
+}
+
+func (c *franzConsumer) consumeLoop(ctx context.Context) {
+	defer close(c.consumerClosed)
+
+	for {
+		select {
+		case <-ctx.Done():
+			c.settings.Logger.Info("Consumer loop stopped due to shutdown")
+			return
+		case <-c.closing:
+			c.settings.Logger.Info("Consumer loop stopped due to shutdown")
+			return
+		default:
+		}
+		// Consume messages until the ctx is cancelled (the client is closed).
+		// NOTE(marclop) we should make the fetch size configurable. It returns
+		// all the internally buffered records. This isn't something that's
+		// configurable in Sarama, and theoretically the max records to iterate
+		// on is a factor of default / max (byte) fetch size.
+		if !c.consume(ctx, -1) {
+			return
+		}
+	}
+}
+
+// consume consumes a batch of messages from the Kafka topic. This is meant to
+// be called in a loop until consume returns false.
+func (c *franzConsumer) consume(ctx context.Context, size int) bool {
+	fetch := c.client.PollRecords(ctx, size)
+	defer c.client.AllowRebalance()
+
+	if err := fetch.Err0(); fetch.IsClientClosed() {
+		c.settings.Logger.Info("consumer stopped", zap.Error(err))
+		return false // Shut down the consumer loop.
+	}
+	// There's a variety of errors that are returned by fetch.Errors(). We
+	// handle the errors that require a client restart above. The rest can
+	// simply be logged and keep fetching.
+	var hasError bool
+	fetch.EachError(func(topic string, partition int32, err error) {
+		c.settings.Logger.Error("consumer fetch error", zap.Error(err),
+			zap.String(attrTopic, topic),
+			zap.String(attrPartition, strconv.Itoa(int(partition))),
+		)
+		if !hasError {
+			hasError = true
+		}
+	})
+	if hasError || fetch.Empty() {
+		return true // Return right away after errors or empty fetch.
+	}
+
+	select { // Check if Shutdown's been called.
+	case <-ctx.Done():
+		return false
+	case <-c.closing:
+		// The client is now closed (Shutdown has been called), so we need to
+		// return and stop consuming. At this point it's safe to return, since
+		// no offsets have been marked for commits (autocommit) or committed.
+		return false
+	default:
+	}
+
```

Comment:
```suggestion
```

Do we need it? IIANM the `fetch.IsClientClosed()` path would have triggered on shutdown. This is a non-blocking select, so it doesn't seem to be adding much.